File: src/main/java/org/snipsniper/configwindow/textpreviewwindow/TextPreviewWindow.java
package org.snipsniper.configwindow.textpreviewwindow;

import org.snipsniper.LangManager;
import org.snipsniper.utils.IClosable;
import org.snipsniper.utils.IFunction;

import javax.swing.*;
import javax.swing.event.DocumentEvent;
import javax.swing.event.DocumentListener;
import java.awt.*;
import java.awt.event.ComponentAdapter;
import java.awt.event.ComponentEvent;
import java.awt.event.WindowAdapter;
import java.awt.event.WindowEvent;
import java.awt.image.BufferedImage;

public class TextPreviewWindow extends JFrame implements IClosable {
    private String text;
    private JTextField input;
    private final JPanel renderer;
    private final JButton saveButton = new JButton(LangManager.getItem("config_label_save"));
    private final JLabel explanationLabel = new JLabel("%hour%, %minute%, %second%, %day%, %month%, %year%, %random%");
    private IFunction onSave;

    public TextPreviewWindow(String title, String text, JPanel renderPanel, BufferedImage icon, JFrame parent, String explanation) {
        this.text = text;
        setDefaultCloseOperation(DO_NOTHING_ON_CLOSE);
        setTitle(title);
        this.explanationLabel.setText(explanation);
        renderer = renderPanel;
        addWindowListener(new WindowAdapter() {
            @Override
            public void windowClosing(WindowEvent e) {
                close();
            }
        });
        addComponentListener(new ComponentAdapter() {
            public void componentResized(ComponentEvent componentEvent) {
                Dimension newSize = new Dimension(getRootPane().getWidth(), getContentPane().getHeight() - input.getHeight() - saveButton.getHeight() - explanationLabel.getHeight());
                renderer.setPreferredSize(newSize);
                renderer.setMinimumSize(newSize);
                renderer.revalidate();
            }
        });
        setSize(256, 256);
        setIconImage(icon);
        setupUI();
        setVisible(true);
        requestFocus();
        pack();
        int x = (int) (parent.getLocation().getX() + parent.getWidth() / 2) - getWidth() / 2;
        int y = (int) (parent.getLocation().getY() + parent.getHeight() / 2) - getHeight() / 2;
        setLocation(x, y);
    }

    private void setupUI() {
        JPanel content = new JPanel(new GridBagLayout());
        GridBagConstraints gbc = new GridBagConstraints();
        gbc.fill = GridBagConstraints.BOTH;
        gbc.gridy = 0;
        content.add(renderer, gbc);
        gbc.gridy = 1;
        input = new JTextField(text);
        input.getDocument().addDocumentListener(new DocumentListener() {
            @Override
            public void insertUpdate(DocumentEvent documentEvent) {
                update();
            }

            @Override
            public void removeUpdate(DocumentEvent documentEvent) {
                update();
            }

            @Override
            public void changedUpdate(DocumentEvent documentEvent) {
                update();
            }

            public void update() {
                text = input.getText();
                renderer.repaint();
            }
        });
        content.add(input, gbc);
        gbc.gridy = 2;
        content.add(explanationLabel, gbc);
        gbc.gridy = 3;
        gbc.fill = GridBagConstraints.VERTICAL;
        saveButton.addActionListener(e -> {
            if(onSave != null) onSave.run();
            dispose();
        });
        content.add(saveButton, gbc);

        add(content);
    }

    public String getText() {
        return text;
    }

    public void setOnSave(IFunction onSave) {
        this.onSave = onSave;
    }

    @Override
    public void close() {
        if(onSave != null)
            onSave.run();
        dispose();
    }
}

File: src/main/kotlin/net/snipsniper/configwindow/textpreviewwindow/TextPreviewWindow.kt
package net.snipsniper.configwindow.textpreviewwindow

import net.snipsniper.utils.IFunction
import net.snipsniper.utils.translate
import org.capturecoop.ccutils.utils.CCIClosable
import java.awt.Dimension
import java.awt.GridBagConstraints
import java.awt.GridBagLayout
import java.awt.event.ComponentAdapter
import java.awt.event.ComponentEvent
import java.awt.event.WindowAdapter
import java.awt.event.WindowEvent
import java.awt.image.BufferedImage
import javax.swing.*
import javax.swing.event.DocumentEvent
import javax.swing.event.DocumentListener

class TextPreviewWindow(wndTitle: String, var text: String, private val renderPanel: JPanel, icon: BufferedImage, parent: JFrame, explanation: String): JFrame(), CCIClosable {
    private val input = JTextField(text)
    private val saveButton = JButton("config_label_save".translate())
    private val explanationLabel = JLabel("%hour%, %minute%, %second%, %day%, %month%, %year%, %random%")
    var onSave: IFunction? = null

    init {
        defaultCloseOperation = DO_NOTHING_ON_CLOSE
        title = wndTitle
        explanationLabel.text = explanation
        addWindowListener(object: WindowAdapter() {
            override fun windowClosing(e: WindowEvent) = close()
        })
        addComponentListener(object: ComponentAdapter() {
            override fun componentResized(e: ComponentEvent) {
                Dimension(rootPane.width, contentPane.height - input.height - saveButton.height - explanationLabel.height).also {
                    renderPanel.preferredSize = it
                    renderPanel.minimumSize = it
                    renderPanel.revalidate()
                }
            }
        })
        setSize(256, 256)
        iconImage = icon
        setupUI()
        isVisible = true
        requestFocus()
        pack()
        fun m(l: Int, p: Int, s: Int) = (l + p / 2) - s / 2
        setLocation(m(parent.location.x, parent.width, width), m(parent.location.y, parent.height, height))
    }

    private fun setupUI() {
        JPanel(GridBagLayout()).also { content ->
            val gbc = GridBagConstraints()
            gbc.fill = GridBagConstraints.BOTH
            gbc.gridy = 0
            content.add(renderPanel, gbc)
            gbc.gridy = 1
            input.document.addDocumentListener(object: DocumentListener {
                override fun insertUpdate(e: DocumentEvent) = update()
                override fun removeUpdate(e: DocumentEvent) = update()
                override fun changedUpdate(e: DocumentEvent) = update()

                fun update() {
                    text = input.text
                    renderPanel.repaint()
                }
            })
            content.add(input, gbc)
            gbc.gridy = 2
            content.add(explanationLabel, gbc)
            gbc.gridy = 3
            gbc.fill = GridBagConstraints.VERTICAL
            saveButton.addActionListener {
                onSave?.run()
                dispose()
            }
            content.add(saveButton, gbc)
            add(content)
        }
    }

    override fun close() {
        onSave?.run()
        dispose()
    }
}